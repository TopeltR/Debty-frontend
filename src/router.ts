import Vue from 'vue';
import Router from 'vue-router';

Vue.use(Router);

export default new Router({
  routes: [
    {
      path: '/',
      name: 'LandingPage',
      component: () => import(/* webpackChunkName: "about" */ './views/LandingPage.vue'),
    },
    {
      path: '/home',
      name: 'home',
      component: () => import(/* webpackChunkName: "about" */ './views/Home.vue'),
    },
    {
      path: '/register',
      name: 'register',
      // route level code-splitting
      // this generates a separate chunk (about.[hash].js) for this route
      // which is lazy-loaded when the route is visited.
      component: () => import(/* webpackChunkName: "about" */ './views/Register.vue'),
    },
    {
      path: '/event/create',
      name: 'create-event',
      component: () => import('./components/CreateEvent.vue'),
    },
    {
      path:'/events' ,
      name: 'events' ,
      component: () => import(/* webpackChunkName: "about" */ './views/Events.vue'),
    },
  ],
});
