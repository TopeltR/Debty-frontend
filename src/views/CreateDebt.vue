<template>
    <div>
        <navbar></navbar>
        <background>
            <b-row>
                <b-col cols="10" offset="1" md="6" offset-md="3" class="w-100">
                    <b-row class="mt-5">
                        <h1 class="w-auto">Create debt</h1>
                        <font-awesome-icon icon="info-circle" class="ml-1"></font-awesome-icon>
                        <b-tooltip target="info"
                                   title="Give your debt an informative title, so that your contact can understand it."
                                   placement="bottom"></b-tooltip>
                    </b-row>
                    <b-row>
                        <form @submit.prevent="saveDebt()" class='w-100'>
                            <div class='form-group pt-3'>
                                <input-label :required="true" for='title'>Title</input-label>
                                <input type='text' class='form-control' v-model='formData.title' id='title'
                                       placeholder='Enter debt title' required minlength="1" maxlength="255">
                            </div>
                            <div class='form-group'>
                                <b-row>
                                    <b-col cols="10" md="11">
                                        <b-row>
                                            <b-col cols="2">
                                                <input-label :required="true">From</input-label>
                                            </b-col>
                                            <b-col cols="10">
                                                <input v-if="!userIsReceiver" type='text' class='form-control'
                                                       v-model='userName' disabled required maxlength="255">
                                                <autocomplete v-else id='payer' v-model='other'
                                                              :placeholder='"Name"'
                                                              :field='field'
                                                              :items='contacts'
                                                              :required='true'
                                                              :key-extractor='getUserFullName'></autocomplete>
                                            </b-col>
                                        </b-row>
                                        <b-row>
                                            <b-col cols="2" class="mt-3">
                                                <input-label :required="true">To</input-label>
                                            </b-col>
                                            <b-col cols="10" class="mt-3">
                                                <input v-if="userIsReceiver" type='text' class='form-control'
                                                       v-model='userName' disabled required maxlength="255">
                                                <autocomplete v-else id='receiver' v-model='other'
                                                              :placeholder='"Name"'
                                                              :field='field'
                                                              :items='contacts'
                                                              :required='true'
                                                              :key-extractor='getUserFullName'></autocomplete>
                                            </b-col>
                                        </b-row>
                                    </b-col>
                                    <b-col cols="2" md="1" v-on:click="switchDebtorReceiver">
                                        <font-awesome-icon icon='arrow-up'
                                                           class="mt-4 p-0 limegreen"></font-awesome-icon>
                                        <font-awesome-icon icon='arrow-down'
                                                           class="mb-2 ml-2 limegreen"></font-awesome-icon>
                                    </b-col>
                                </b-row>
                            </div>
                            <div class='form-group'>
                                <b-row>
                                    <b-col cols='2'>
                                        <input-label :required="true" for='sum' class='mt-2'>Sum</input-label>
                                    </b-col>
                                    <b-col cols='3'>
                                        <input id='sum' type='number' class='form-control pr-0'
                                               v-model='formData.sum'
                                               placeholder='0' required min="0" maxlength="255">
                                    </b-col>
                                    <b-col cols='1' class='mt-2 pl-0'>â‚¬</b-col>
                                </b-row>
                            </div>
                            <b-row class='mt-4 text-center'>
                                <b-col>
                                    <b-btn type="submit" class="w-100" variant="primary">Save</b-btn>
                                </b-col>
                            </b-row>
                        </form>
                    </b-row>
                </b-col>
            </b-row>
        </background>
    </div>
</template>

<script>
    import router from '@/router.ts';
    import Background from '@/components/Background.vue';
    import Navbar from '@/components/Navbar.vue';
    import Autocomplete from '@/components/Autocomplete.vue';
    import Spinner from '@/components/Spinner.vue';
    import userStore from '@/stores/UserStore.ts';
    import InputLabel from "@/components/InputLabel.vue";


    export default {
        name: 'CreateDebt',
        components: {InputLabel, Background, Navbar, Autocomplete, Spinner},
        data: () => ({
            formData: {
                title: '',
                payer: null,
                receiver: null,
                owner: null,
                sum: null,
            },
            field: {value: ''},
            contacts: [],
            user: null,
            other: null,
            userIsReceiver: true,
            userName: null,
        }),
        async mounted() {
            this.user = await userStore.getUser();
            this.formData.receiver = this.user;
            this.formData.owner = this.user;
            this.userName = this.getUserFullName(this.user);

            const response = await this.$http.get('/contacts/' + this.user.id + '/users');
            this.contacts = response.data.filter((u) => u.email !== this.user.email);
        },
        methods: {
            getUserFullName(user) {
                return user.lastName == null ? user.firstName : user.firstName + ' ' + user.lastName;
            },
            async saveDebt() {
                const result = await this.$http.post('/debts', this.formData);
                router.push('/debts/' + result.data.id);
            },
            switchDebtorReceiver() {
                this.userIsReceiver = !this.userIsReceiver;
                if (this.userIsReceiver) {
                    this.formData.receiver = this.user;
                    this.formData.payer = this.other;
                } else {
                    this.formData.receiver = this.other;
                    this.formData.payer = this.user;
                }
            },
        },
    };
</script>

<style scoped>
    .limegreen {
        color: limegreen !important;
    }
</style>
