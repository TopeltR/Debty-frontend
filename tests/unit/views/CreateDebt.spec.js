import CreateDebt from '@/views/CreateDebt';
import userStore from '@/stores/UserStore'
import {mount} from "@vue/test-utils";
import Vue from "vue";
import BootstrapVue from 'bootstrap-vue';
import router from '@/router.ts';

describe('CreateDebt', () => {
    console.error = jest.fn();
    console.warn = jest.fn();
    Vue.use(BootstrapVue);
    const get = jest.fn(() => Promise.resolve({
        data: [
            {id: 2, email: 'user2@email.com'},
            {id: 1, email: 'user1@email.com'}
        ]
    }));
    const post = jest.fn(() => Promise.resolve({data: {id: 5}}));
    Vue.prototype.$http = {get, post};

    const user = {id: 1, firstName: 'first', email: 'user1@email.com'};
    userStore.getUser = jest.fn(() => Promise.resolve(user));

    describe('mounted', () => {
        it('should set formData receiver and owner and user, userName', async () => {
            const createDebt = mount(CreateDebt);

            await Vue.nextTick();

            expect(createDebt.vm.formData.receiver).toEqual(user);
            expect(createDebt.vm.formData.owner).toEqual(user);
            expect(createDebt.vm.user).toEqual(user);
            expect(createDebt.vm.userName).toEqual('first');
        });

        it('should request user contacts', async () => {
            const createDebt = mount(CreateDebt);

            await Vue.nextTick();
            await Vue.nextTick();

            expect(get).toHaveBeenCalledWith('/contact/id/1');
            expect(createDebt.vm.contacts).toEqual([{id: 2, email: 'user2@email.com'}])
        })
    });

    describe('createDebt', () => {
        it('should post debt and redirect to debts page', async () => {
            router.push = jest.fn();
            const createDebt = mount(CreateDebt);
            createDebt.vm.formData.payer = user;

            await createDebt.vm.createDebt();

            expect(post).toHaveBeenCalledWith('/debts', {
                owner: {
                    email: "user1@email.com",
                    firstName: "first",
                    id: 1
                },
                payer: {email: "user1@email.com", firstName: "first", id: 1},
                receiver: {email: "user1@email.com", firstName: "first", id: 1},
                sum: null,
                title: ""
            });
            expect(router.push).toHaveBeenCalledWith('/debts/5');
        })
    });

    it('should render', () => {
        const createDebt = mount(CreateDebt);

        expect(createDebt.html()).toMatchSnapshot();
    })
});
