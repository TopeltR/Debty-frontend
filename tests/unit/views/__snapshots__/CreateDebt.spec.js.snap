// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateDebt should render 1`] = `
<div>
  <div class="container-fluid">
    <div class="row PB65px">
      <nav class="navbar fixed-top navbar-light navbar-expand-sm" style="background-color: rgb(232, 232, 232); width: 100%;"><button type="button" aria-label="Toggle navigation" aria-controls="nav_collapse" aria-expanded="false" class="navbar-toggler"><span class="navbar-toggler-icon"></span></button>
        <div class="navbar-brand limegreen brand">debty</div>
        <div id="nav_collapse" class="navbar-collapse collapse" style="display: none;">
          <ul class="navbar-nav">
            <li class="nav-item"><a href="/home" target="_self" class="nav-link">
                Home
              </a></li>
          </ul>
          <ul class="navbar-nav">
            <li class="nav-item"><a href="/events" target="_self" class="nav-link">
                Events
              </a></li>
          </ul>
          <ul class="navbar-nav">
            <li class="nav-item"><a href="/debts" target="_self" class="nav-link">
                Debts
              </a></li>
          </ul>
          <ul class="navbar-nav">
            <li class="nav-item"><a href="/contacts" target="_self" class="nav-link">
                Contacts
                <!----></a></li>
          </ul>
          <ul class="navbar-nav">
            <li class="nav-item"><a href="/profile" target="_self" class="nav-link">
                Profile
              </a></li>
          </ul>
        </div>
      </nav>
    </div>
  </div>
  <div class="diagonal-background container-fluid">
    <div class="container">
      <div class="row">
        <div class="mt-5 pt-3 col-sm-12 col-md-6 offset-md-3">
          <div class="row">
            <h2>Create debt</h2>
            <font-awesome-icon id="info" icon="info-circle" class="ml-1"></font-awesome-icon>
            <div aria-hidden="true" class="d-none" style="display: none;">
              <div></div>
            </div>
            <form class="wide">
              <div class="form-group"><label for="title">Title:</label> <input type="text" id="title" placeholder="Enter debt title" class="form-control"></div>
              <div class="form-group">
                <div class="row">
                  <div class="col-11">
                    <div class="row">
                      <div class="col-2"><label>From:</label></div>
                      <div class="col-10">
                        <div class="autocomplete" id="payer"><input type="text" placeholder="Name" class="form-control">
                          <ul id="autocomplete-results" class="autocomplete-results" style="display: none;"></ul>
                        </div>
                      </div>
                    </div>
                    <div class="row">
                      <div class="mt-3 col-2"><label>To:</label></div>
                      <div class="mt-3 col-10"><input type="text" disabled="disabled" class="form-control"></div>
                    </div>
                  </div>
                  <div class="col-1">
                    <font-awesome-icon icon="arrow-up" class="mt-4 p-0 limegreen"></font-awesome-icon>
                    <font-awesome-icon icon="arrow-down" class="mb-2 ml-2 limegreen"></font-awesome-icon>
                  </div>
                </div>
              </div>
              <div class="form-group">
                <div class="row">
                  <div class="col-2"><label for="sum" class="mt-1">Sum:</label></div>
                  <div class="col-md-2 col-3"><input id="sum" type="text" placeholder="0" class="form-control pr-0"></div>
                  <div class="pl-0 mt-1 col-1">â‚¬</div>
                </div>
              </div>
              <div class="row mt-4">
                <div class="col-sm-12"><button type="button" class="btn btn-primary wide">
                    Create debt
                  </button></div>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
